api/kubernetes/client.go

```diff
@@ -0,0 +1,169 @@
+func (c client) DownscaleWorkload(replicas int, workload scalable.Workload, ctx context.Context) error {
+	originalReplicas, err := workload.GetCurrentReplicas()
+	if err != nil {
+		return fmt.Errorf("failed to get original replicas for workload: %w", err)
+	}
+	if originalReplicas == replicas {
+		slog.Debug("workload is already at downscale replicas, skipping", "workload", workload.GetName(), "namespace", workload.GetNamespace())
+		return nil
+	}
+
+	workload.SetReplicas(replicas)
+	err = c.setOriginalReplicas(originalReplicas, workload)
+	if err != nil {
+		return fmt.Errorf("failed to set original replicas annotation: %w", err)
+	}
+	err = workload.Update(c.clientset, ctx)
+	if err != nil {
+		return fmt.Errorf("failed to update workload: %w", err)
+	}
+	slog.Debug("successfully scaled down workload", "workload", workload.GetName(), "namespace", workload.GetNamespace())
+	return nil
+}
+
+// UpscaleWorkload upscales the workload to the original replicas
+func (c client) UpscaleWorkload(workload scalable.Workload, ctx context.Context) error {
+	currentReplicas, err := workload.GetCurrentReplicas()
+	if err != nil {
+		return fmt.Errorf("failed to get original replicas for workload: %w", err)
+	}
+	originalReplicas, err := c.getOriginalReplicas(workload)
+	if err != nil {
+		return fmt.Errorf("failed to get original replicas for workload: %w", err)
+	}
+	if originalReplicas == values.Undefined {
+		slog.Debug("original replicas is not set, skipping", "workload", workload.GetName(), "namespace", workload.GetNamespace())
+		return nil
+	}
+	if originalReplicas == currentReplicas {
+		slog.Debug("workload is already at original replicas, skipping", "workload", workload.GetName(), "namespace", workload.GetNamespace())
+		return nil
+	}
+
+	workload.SetReplicas(originalReplicas)
+	err = c.removeOriginalReplicas(workload)
+	if err != nil {
+		return fmt.Errorf("failed to set original replicas annotation: %w", err)
+	}
+	err = workload.Update(c.clientset, ctx)
+	if err != nil {
+		return fmt.Errorf("failed to update workload: %w", err)
+	}
+	slog.Debug("successfully scaled up workload", "workload", workload.GetName(), "namespace", workload.GetNamespace())
+	return nil
+}
+
+// setOriginalReplicas sets the original replicas annotation on the workload
+func (c client) setOriginalReplicas(originalReplicas int, workload scalable.Workload) error {
+	annotations := workload.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	annotations[annotationOriginalReplicas] = fmt.Sprintf("%d", originalReplicas)
+	workload.SetAnnotations(annotations)
+	return nil
+}
+
+// getOriginalReplicas gets the original replicas annotation on the workload. nil is undefined
+func (c client) getOriginalReplicas(workload scalable.Workload) (int, error) {
+	annotations := workload.GetAnnotations()
+	originalReplicasString, ok := annotations[annotationOriginalReplicas]
+	if !ok {
+		return values.Undefined, nil
```

Comment:
This is intentional, because it being unset isn't an "error" as it is a valid state of original replicas. There is a if check for it being "Undefined" where its used that also just "skips" it because of this.